lib/trino-filesystem/src/main/java/io/trino/filesystem/Location.java

```diff
@@ -178,6 +178,17 @@ public String fileName()
         return path.substring(path.lastIndexOf('/') + 1);
     }
 
+    /**
+     * Returns a new location with the same parent directory as the current location,
+     * but with the filename corresponding to the specified `filename`.
+     * The location must be a valid file location.
+     */
+    public Location replaceFilename(String filename)
+    {
+        verifyValidFileLocation();
+        return Location.of(location.replaceAll("[^/]+$", "") + filename);
```

Comment:
- use lastIndexOf("/") instead of regular expressions to be much faster 
- use `io.trino.filesystem.Location#withPath` to avoid re-parsing the location